src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -274,19 +276,48 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
             {
                 Requires.NotNull(operation, nameof(operation));
 
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons in the JTF enabled scenario, we want to minimize the number of Joins performed, and also
+                // keep the size of the JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a
+                // JTF.RunAsync. This requires us to not ConfigureAwait(true) on the semaphore. However, that prevents us from
+                // resuming on the correct sync context. To partially fix this, we will at least resume you on the main thread or
+                // thread pool.
+                AsyncSemaphore.Releaser releaser;
+                bool resumeOnMainThread = this.joinableTaskCollection?.Context?.IsOnMainThread == true;
+                bool usingJtf = this.joinableTaskFactory != null;
+                using (this.joinableTaskCollection?.Join())
+                {
+                    // Use ConfiguredAwaitRunInline() as ConfigureAwait(true) will deadlock due to not being inside a JTF.RunAsync().
+                    releaser = usingJtf
+                        ? await this.semaphore.EnterAsync(cancellationToken).ConfigureAwaitRunInline()
+                        : await this.semaphore.EnterAsync(cancellationToken).ConfigureAwait(true);
+                }
+
                 await this.ExecuteCoreAsync(async delegate
                 {
-                    using (this.joinableTaskCollection?.Join())
+                    try
                     {
-                        var releaser = await this.semaphore.EnterAsync(cancellationToken).ConfigureAwait(true);
-                        try
-                        {
-                            await operation().ConfigureAwaitRunInline();
-                        }
-                        finally
+                        if (usingJtf)
                         {
-                            DisposeReleaserNoThrow(releaser);
+                            if (resumeOnMainThread)
+                            {
+                                // Return to the main thread if we started there.
+                                await this.joinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                            }
+                            else
+                            {
+                                await TaskScheduler.Default;
+                            }
+
+                            // Always yield to prevent running on the stack that released the semaphore.
+                            await Task.Yield();
                         }
```

Comment:
We're so nearly there. But I just realized that this change introduces a yield even when there was no semaphore contention for the caller (we should have a test for this).
If you revise your `release = usingJtf` assignment. to *not* await directly, but simply assign the unawaited result to a local variable, then we can call `IsCompleted` on that ourselves and save another local indicating whether it was completed (telling us there was no contention and thus no yielding). We then of course will need to await that variable that stores the `EnterAsync` result. We can then condition *this* `await Task.Yield`  based on whether `!IsCompleted` from earlier.